src/jsTyping/jsTyping.ts

```diff
@@ -214,17 +204,99 @@ namespace ts.JsTyping {
+            // `npm@>=3 <7`.
+
+            // Package names that do **not** provide their own typings, so
+            // we'll look them up.
+            const packageNames: string[] = [];
+
+            const dependencyManifestNames = manifestTypingNames
+                // This is #1 described above.
+                ? manifestTypingNames.map(typingName => combinePaths(packagesFolderPath, typingName, manifestName))
+                // And #2. Depth = 3 because scoped packages look like `node_modules/@foo/bar/package.json`
+                : host.readDirectory(packagesFolderPath, [Extension.Json], /*excludes*/ undefined, /*includes*/ undefined, /*depth*/ 3)
+                    .filter(manifestPath => {
+                        if (getBaseFileName(manifestPath) !== manifestName) {
+                            return false;
+                        }
+                        // It's ok to treat
+                        // `node_modules/@foo/bar/package.json` as a manifest,
+                        // but not `node_modules/jquery/nested/package.json`.
+                        // We only assume depth 3 is ok for formally scoped
+                        // packages. So that needs this dance here.
+                        const pathComponents = getPathComponents(normalizePath(manifestPath));
+                        const isScoped = pathComponents[pathComponents.length - 3][0] === "@";
+                        return isScoped && pathComponents[pathComponents.length - 4].toLowerCase() === modulesDirName || // `node_modules/@foo/bar`
+                            !isScoped && pathComponents[pathComponents.length - 3].toLowerCase() === modulesDirName; // `node_modules/foo`
+                    });
+
+            if (log) log(`Searching for typing names in ${packagesFolderPath}; all files: ${JSON.stringify(dependencyManifestNames)}`);
+
+            // Once we have the names of things to look up, we iterate over
+            // and either collect their included typings, or add them to the
+            // list of typings we need to look up separately.
+            for (const manifestPath of dependencyManifestNames) {
+                const normalizedFileName = normalizePath(manifestPath);
+                const result = readConfigFile(normalizedFileName, (path: string) => host.readFile(path));
+                const manifest: PackageJson = result.config;
+
+                // If the package has its own d.ts typings, those will take precedence. Otherwise the package name will be used
+                // to download d.ts files from DefinitelyTyped
+                if (!manifest.name) {
+                    continue;
+                }
+                const ownTypes = manifest.types || manifest.typings;
+                if (ownTypes) {
+                    const absolutePath = getNormalizedAbsolutePath(ownTypes, getDirectoryPath(normalizedFileName));
+                    if (log) log(`    Package '${manifest.name}' provides its own types.`);

```

Comment:
I'm sure @amcasey wouldn't be too mad if you incorporated his changes from #46877 into here?

Reply:
I'm fine with that. Better than making him deal with the conflict :)